src/cargo/util/config.rs

```diff
@@ -32,22 +33,119 @@ use util::Filesystem;
 
 use self::ConfigValue as CV;
 
+#[derive(Clone, Debug)]
+pub struct CargoDirs {
+    pub home_dir: Filesystem,
+    pub current_dir: PathBuf,
+    pub cache_dir: Filesystem,
+    pub config_dir: Filesystem,
+    pub data_dir: PathBuf,
+    pub bin_dir: PathBuf,
+}
+
+impl CargoDirs {
+    /// Computes the paths to directories used by cargo to retrieve and store
+    /// cache, config, ... files.
+    // This is written in the most straight-forward way possible, because it is
+    // hard as-is to understand all the different options, without trying to
+    // save lines of code.
+    pub fn new() -> CargoResult<CargoDirs> {
+        let home_dir = ::home::home_dir().ok_or_else(|| {
+            format_err!("Cargo couldn't find your home directory. \
+                         This probably means that $HOME was not set.")
+        })?;
```

Comment:
So looks like we don't use home_dir for anything except fallback, so let's move this as far down as possible, so that we don't actually fail if `HOME` is not set, but explicit directories are! We might want to test this behavior as well: working without home directory is great for reproducible and isolated builds.